### PR TITLE
[RLlib] PPO training iteration fn

### PR DESCRIPTION
**Pull request by sven1977**
Move PPOTrainer into new training iteration fn API
* Set _disable_execution_plan_api=True for PPO

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/agents/ppo/ppo.py [hunk 1]
```diff
@@ -410,6 +418,86 @@ def get_default_policy_class(self, config: TrainerConfigDict) -> Type[Policy]:
...
+            new_sample_batches = synchronous_parallel_sample(self.workers)
+            sample_batches.extend(new_sample_batches)
+            num_env_steps += sum(len(s) for s in new_sample_batches)
+            num_agent_steps += sum(s.agent_steps() for s in new_sample_batches)
+        self._counters[NUM_ENV_STEPS_SAMPLED] += num_env_steps
+        self._counters[NUM_AGENT_STEPS_SAMPLED] += num_agent_steps
+
+        # Combine all batches at once and treat as multi-agent from here on.
+        train_batch = SampleBatch.concat_samples(sample_batches)
+        train_batch = train_batch.as_multi_agent()
```
Comment (gjoliver) [inline rllib/agents/ppo/ppo.py:442]: not really related, but it feels weird to me to have every agent do this as_multi_agent() call everywhere.

File: rllib/agents/ppo/ppo.py [hunk 2]
```diff
@@ -410,6 +418,86 @@ def get_default_policy_class(self, config: TrainerConfigDict) -> Type[Policy]:
...
+            if Postprocessing.ADVANTAGES not in batch:
+                raise KeyError(
+                    f"`{Postprocessing.ADVANTAGES}` not found in SampleBatch for "
+                    f"policy `{policy_id}`! Maybe this policy fails to add "
+                    f"`{Postprocessing.ADVANTAGES}` in its `postprocess_trajectory` "
+                    f"method? Or this policy is not meant to learn at all and you "
+                    "forgot to remove it from the list under `config."
+                    "multiagent.policies_to_train`."
+                )
+            batch[Postprocessing.ADVANTAGES] = standardized(
```
Comment (gjoliver) [inline rllib/agents/ppo/ppo.py:455]: why do we need this here, instead of in the policy's postprocessing function?

File: rllib/agents/ppo/ppo.py [hunk 3]
```diff
@@ -410,6 +418,86 @@ def get_default_policy_class(self, config: TrainerConfigDict) -> Type[Policy]:
...
+
+        # Update KL on all trainable policies within the local (trainer)
+        # Worker.
+        self.workers.local_worker().foreach_policy_to_train(update)
+
+        # Update global vars on .
+        self.workers.local_worker().set_global_vars(global_vars)
+
+        # Warn about bad reward scales and return training metrics.
+        #warn_about_bad_reward_scales(self.config, train_results)
```
Comment (gjoliver) [inline rllib/agents/ppo/ppo.py:497]: delete? why have commented out lines here?

**Review by gjoliver — COMMENTED** [review]
looks mostly reasonable.
I wonder what do we have this and https://github.com/ray-project/ray/pull/23673?

File: rllib/agents/ppo/ppo.py (hunk 2 above)
Comment (sven1977) [reply rllib/agents/ppo/ppo.py:455]: I guess to standardize across the entire train batch, and not necessarily only across each (possibly truncated) episode trajectory.

File: rllib/agents/ppo/ppo.py (hunk 3 above)
Comment (sven1977) [reply rllib/agents/ppo/ppo.py:497]: Yeah, we got to fix this first, sorry.
Metrics from rollouts are pulled after(!) returning from `training_iteration`, so having this here won't work. We could add the check at the beginning of the iteration fn and run the analysis (only needs episode_reward_mean and vf_loss) on the previous iteration's data. ... Let me fix this.
I also noticed that we don't perform these checks in the multi-agent case, so we can add this here as well.

File: rllib/agents/ppo/ppo.py (hunk 1 above)
Comment (sven1977) [reply rllib/agents/ppo/ppo.py:442]: I feel like for the entire ["collect batches until train_batch_size reached"]-problem, we should have a utility function. It all comes from the fact that we return different things from `sample()`, depending on whether the setup is multi-agent or not. I feel like we should always return MultiAgentBatch from `sample()`, never just SampleBatch.

**Comment by sven1977** [comment]
Closing this due to (almost) identity to https://github.com/ray-project/ray/pull/23673

**Pull request closed without merging**